apiserver/Makefile

```diff
@@ -43,16 +16,81 @@ jupyter_notebook_config.py:
-$(eval $(call DOCKER_IMAGE,apiserver,install-hail-locally jupyter_notebook_config.py,APISERVER_IMAGE))
-$(eval $(call DOCKER_IMAGE,hail-jupyter,install-hail-locally jupyter_notebook_config.py,HAIL_JUPYTER_IMAGE))
+build-common: install-hail-locally jupyter_notebook_config.py
+
+ifeq ($(IN_HAIL_CI),1)
+build-spark-base: build-common
+	docker pull debian:9.5
+	-docker pull gcr.io/$(PROJECT)/spark-base:latest
+	docker build -t spark-base -f Dockerfile.spark-base . --cache-from spark-base,gcr.io/$(PROJECT)/spark-base:latest,debian:9.5
+
+build-hail-base: build-common
+	-docker pull gcr.io/$(PROJECT)/spark-base:latest
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	docker build -t hail-base -f Dockerfile.hail-base . --cache-from hail-base,gcr.io/$(PROJECT)/hail-base:latest,spark-base
+
+build-spark-master: build-common
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/spark-master:latest
+	docker build -t spark-master -f Dockerfile.spark-master . --cache-from spark-master,gcr.io/$(PROJECT)/spark-master:latest,hail-base
+
+build-spark-worker: build-common
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/spark-worker:latest
+	docker build -t spark-worker -f Dockerfile.spark-worker . --cache-from spark-worker,gcr.io/$(PROJECT)/spark-worker:latest,hail-base
+
+build-apiserver: build-common
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/apiserver:latest
+	docker build -t apiserver -f Dockerfile.apiserver . --cache-from apiserver,gcr.io/$(PROJECT)/apiserver:latest,hail-base
+
+build-hail-jupyter: build-common
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/hail-jupyter:latest
+	docker build -t hail-jupyter -f Dockerfile.hail-jupyter . --cache-from hail-jupyter,gcr.io/$(PROJECT)/hail-jupyter:latest,hail-base
 
 build: build-spark-base build-hail-base build-spark-master build-spark-worker build-apiserver build-hail-jupyter
 
-push: push-spark-base push-hail-base push-spark-master push-spark-worker push-apiserver push-hail-jupyter
+SPARK_MASTER_IMAGE="gcr.io/$(PROJECT)/spark-master:$(shell docker images -q --no-trunc spark-master | sed -e 's,[^:]*:,,')"
+SPARK_WORKER_IMAGE="gcr.io/$(PROJECT)/spark-worker:$(shell docker images -q --no-trunc spark-worker | sed -e 's,[^:]*:,,')"
+APISERVER_IMAGE="gcr.io/$(PROJECT)/apiserver:$(shell docker images -q --no-trunc apiserver | sed -e 's,[^:]*:,,')"
+HAIL_JUPYTER_IMAGE="gcr.io/$(PROJECT)/hail-jupyter:$(shell docker images -q --no-trunc hail-jupyter | sed -e 's,[^:]*:,,')"
+
+push: build
+	docker tag spark-master $(SPARK_MASTER_IMAGE)
+	docker push $(SPARK_MASTER_IMAGE)
+	docker tag spark-worker $(SPARK_WORKER_IMAGE)
+	docker push $(SPARK_WORKER_IMAGE)
+	docker tag apiserver $(APISERVER_IMAGE)
+	docker push $(APISERVER_IMAGE)
+	docker tag hail-jupyter $(HAIL_JUPYTER_IMAGE)
+	docker push $(HAIL_JUPYTER_IMAGE)
```

Comment:
these need to push the `latest` tag as well or we won't get any cache'ing benefit.